File: core/src/main/java/y2k/joyreactor/services/TagsService.java
package y2k.joyreactor.services;

import rx.Observable;
import y2k.joyreactor.Tag;
import y2k.joyreactor.services.repository.MyTagQuery;
import y2k.joyreactor.services.repository.Repository;
import y2k.joyreactor.services.requests.AddTagRequest;
import y2k.joyreactor.services.requests.UserNameRequest;
import y2k.joyreactor.services.synchronizers.MyTagFetcher;

import java.util.List;

/**
 * Created by y2k on 11/24/15.
 */
public class TagsService {

    private Repository<Tag> repository;
    private MyTagFetcher synchronizer;

    public TagsService(Repository<Tag> repository, MyTagFetcher synchronizer) {
        this.repository = repository;
        this.synchronizer = synchronizer;
    }

    public Observable<List<Tag>> getMyTags() {
        return getFromRepo().mergeWith(
                synchronizer.synchronize().flatMap(_void -> getFromRepo()));

    }

    private Observable<List<Tag>> getFromRepo() {
        return repository.queryAsync(new MyTagQuery());
    }

    public Observable<?> addTag(String tag) {
        return new AddTagRequest(tag).request();
    }

    public Observable<Tag> getTagForFavorite() {
        return new UserNameRequest().request().map(Tag::makeFavorite);
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/TagsService.kt
package y2k.joyreactor.services

import rx.Observable
import rx.functions.Func1
import y2k.joyreactor.Tag
import y2k.joyreactor.services.repository.MyTagQuery
import y2k.joyreactor.services.repository.Repository
import y2k.joyreactor.services.requests.AddTagRequest
import y2k.joyreactor.services.requests.UserNameRequest
import y2k.joyreactor.services.synchronizers.MyTagFetcher

/**
 * Created by y2k on 11/24/15.
 */
class TagsService(
        private val repository: Repository<Tag>,
        private val synchronizer: MyTagFetcher) {

    fun getMyTags(): Observable<List<Tag>> {
        return getFromRepo().mergeWith(
                synchronizer.synchronize().flatMap { getFromRepo() })
    }

    private fun getFromRepo(): Observable<List<Tag>> {
        return repository.queryAsync(MyTagQuery())
    }

    fun addTag(tag: String): Observable<*> {
        return AddTagRequest(tag).request()
    }

    fun getTagForFavorite(): Observable<Tag> {
        return UserNameRequest().request().map { Tag.makeFavorite(it) }
    }
}